executor/common_linux.h

```diff
@@ -1435,6 +1435,85 @@ static long syz_io_uring_complete(volatile long a0)
 
 #endif
 
+#if SYZ_EXECUTOR || __NR_syz_io_uring_setup
+
+struct io_sqring_offsets {
+	uint32 head;
+	uint32 tail;
+	uint32 ring_mask;
+	uint32 ring_entries;
+	uint32 flags;
+	uint32 dropped;
+	uint32 array;
+	uint32 resv1;
+	uint64 resv2;
+};
+
+struct io_cqring_offsets {
+	uint32 head;
+	uint32 tail;
+	uint32 ring_mask;
+	uint32 ring_entries;
+	uint32 overflow;
+	uint32 cqes;
+	uint64 resv[2];
+};
+
+struct io_uring_params {
+	uint32 sq_entries;
+	uint32 cq_entries;
+	uint32 flags;
+	uint32 sq_thread_cpu;
+	uint32 sq_thread_idle;
+	uint32 features;
+	uint32 resv[4];
+	struct io_sqring_offsets sq_off;
+	struct io_cqring_offsets cq_off;
+};
+
+#define IORING_OFF_SQ_RING 0
+#define IORING_OFF_SQES 0x10000000ULL
+
+#include <unistd.h>
+
+#ifndef __NR_io_uring_setup
+#define __NR_io_uring_setup 425
+#endif
+
+// Wrapper for io_uring_setup and the subsequent mmap calls that map the ring and the sqes
+static long syz_io_uring_setup(volatile long a0, volatile long a1, volatile long a2, volatile long a3, volatile long a4, volatile long a5)
+{
+	// syzlang: syz_io_uring_setup(entries int32[1:IORING_MAX_ENTRIES], params ptr[inout, io_uring_params], addr_ring vma, addr_sqes vma, ring_ptr ptr[out, ring_ptr], sqes_ptr ptr[out, sqes_ptr]) fd_io_uring
+	// C:       syz_io_uring_setup(uint32 entries, struct io_uring_params* params, void* mmap_addr_ring, void* mmap_addr_sqes, void** ring_ptr_out, void** sqes_ptr_out) // returns uint32 fd_io_uring
+
+	// Cast to original
+	uint32 entries = (uint32)a0;
+	struct io_uring_params* setup_params = (struct io_uring_params*)a1;
+	void* vma1 = (void*)a2;
+	void* vma2 = (void*)a3;
+	void** ring_ptr_out = (void**)a4;
+	void** sqes_ptr_out = (void**)a5;
+
+	uint32 fd_io_uring = syscall(__NR_io_uring_setup, entries, setup_params);
+
+	// Compute the ring sizes
+	uint32 sq_ring_sz = setup_params->sq_off.array + setup_params->sq_entries * sizeof(uint32);
+	uint32 cq_ring_sz = setup_params->cq_off.cqes + setup_params->cq_entries * SIZEOF_IO_URING_CQE;
+
+	// Asssumed IORING_FEAT_SINGLE_MMAP, which is always the case with the current implementation
+	// The implication is that the sq_ring_ptr and the cq_ring_ptr are the same but the
+	// difference is in the offsets to access the fields of these rings.
+	uint32 ring_sz = sq_ring_sz > cq_ring_sz ? sq_ring_sz : cq_ring_sz;
+	*ring_ptr_out = (void*)syscall(__NR_mmap, vma1, ring_sz, 0x3, 0x8011, fd_io_uring, IORING_OFF_SQ_RING);
```

Comment:
I think we use normal libc calls in pseudo-syscalls. So you could just use mmap call.

Reply:
[fixed](https://github.com/google/syzkaller/pull/1984/files#diff-def271e38e744a7dca0b6e9075acb4f2R1511-R1515)

It seems like `io_uring_setup` syscall does not have a libc call. We need to include liburing headers to have the C func ([here](https://elixir.bootlin.com/linux/latest/source/tools/io_uring/liburing.h#L56)).

Instead, I defined the syscall number [here](https://github.com/google/syzkaller/pull/1984/files#diff-def271e38e744a7dca0b6e9075acb4f2R1480-R1486) as in [here](https://elixir.bootlin.com/linux/latest/source/tools/io_uring/syscall.c#L15), and called it with `syscall` ([here](https://github.com/google/syzkaller/pull/1984/files#diff-def271e38e744a7dca0b6e9075acb4f2R1502)).